File: utils/logging.py
"""Application logging configuration helpers.

The project uses Python's standard ``logging`` package behind a small helper so
modules can consistently request namespaced loggers without duplicating handler
configuration.
"""

from __future__ import annotations

import logging
import os
from typing import Final

_DEFAULT_LOGGER_NAME: Final[str] = "admin_panel_dashboard"
_DEFAULT_LOG_FORMAT: Final[str] = (
    "%(asctime)s %(levelname)s [%(name)s] %(message)s"
)
_CONFIGURED_LOGGERS: set[str] = set()


def _get_log_level() -> int:
    """Resolve the configured log level from ``LOG_LEVEL``.

    Invalid values fall back to ``INFO`` so a typo in the environment does not
    break application startup.
    """
    configured_level = os.getenv("LOG_LEVEL", "INFO").upper()
    return getattr(logging, configured_level, logging.INFO)


def get_logger(name: str | None = None) -> logging.Logger:
    """Return a configured project logger.

    Args:
        name: Optional child logger name. When omitted, the root project logger
            named ``admin_panel_dashboard`` is returned.

    Returns:
        A logger configured with one stream handler and a consistent formatter.
    """
    logger_name = _DEFAULT_LOGGER_NAME if not name else f"{_DEFAULT_LOGGER_NAME}.{name}"
    logger = logging.getLogger(logger_name)
    logger.setLevel(_get_log_level())

    if logger_name not in _CONFIGURED_LOGGERS:
        handler = logging.StreamHandler()
        handler.setFormatter(logging.Formatter(_DEFAULT_LOG_FORMAT))
        logger.addHandler(handler)
        logger.propagate = False
        _CONFIGURED_LOGGERS.add(logger_name)

    return logger

File: utils/logging.js
const DEFAULT_LOGGER_NAME = 'admin_panel_dashboard';
const LEVELS = { DEBUG: 10, INFO: 20, WARNING: 30, ERROR: 40 };
const configuredLoggers = new Map();

function currentLevel() {
  return LEVELS[String(process.env.LOG_LEVEL || 'INFO').toUpperCase()] ?? LEVELS.INFO;
}

function shouldLog(level) {
  return LEVELS[level] >= currentLevel();
}

function write(level, loggerName, args) {
  if (!shouldLog(level)) return;
  const timestamp = new Date().toISOString();
  const line = `${timestamp} ${level} [${loggerName}] ${args.map(String).join(' ')}`;
  const output = level === 'ERROR' || level === 'WARNING' ? console.error : console.log;
  output(line);
}

export function getLogger(name = null) {
  const loggerName = name ? `${DEFAULT_LOGGER_NAME}.${name}` : DEFAULT_LOGGER_NAME;
  if (!configuredLoggers.has(loggerName)) {
    configuredLoggers.set(loggerName, {
      name: loggerName,
      handlers: ['console'],
      debug: (...args) => write('DEBUG', loggerName, args),
      info: (...args) => write('INFO', loggerName, args),
      warning: (...args) => write('WARNING', loggerName, args),
      warn: (...args) => write('WARNING', loggerName, args),
      error: (...args) => write('ERROR', loggerName, args),
    });
  }
  return configuredLoggers.get(loggerName);
}
